Attempt at generic factory furniture (consume/produce)

dusho opened the pull request:
Working in generic consume/produce (discussed in #604)
Can do:
- added optional Workshop (FurnitureWorkshop.cs) to Furniture .xml
- furniture can handle multiple production chains (recipes)
- can have multiple input/output slots
- added context menu if there is more than 1 available production chain
- load/save seems to be working (Inventory.locked is not persisted, after load, items from input are taken away - new issue, I guess)

There is new temporary Metal Furnace (that works as Metal Smelter) just to try things out. Idea is to get rid of most of repeated code in LUA for furniture that consumes and produces something
New art for Power cell press, input: 3x uranium, 2x steel plate, output: 1x power cell

vogonistic commented on Assets/Editor/FactoryInfoTest.cs:
```diff
@@ -0,0 +1,48 @@
+        fi.PossibleProductions = new System.Collections.Generic.List<FactoryInfo.ProductionChain>();
+        //fi.InputSlots = new List<FactoryInfo.IntVector2>() { new FactoryInfo.IntVector2(1, 0) };
+        //fi.OutputSlots = new List<FactoryInfo.IntVector2>() { new FactoryInfo.IntVector2(2, 0) };
+        var chain1 = new FactoryInfo.ProductionChain() { Name="Iron smelting", ProcessingTime = 4.0f };
```
This test file doesn't look formatted according to the rules in the CONTRIBUTING.md file. It's pretty hard to read.

vogonistic commented on Assets/Scripts/Models/Furniture.Factory.cs:
```diff
@@ -0,0 +1,185 @@
+
+        if (!string.IsNullOrEmpty(curSetupChainName))
+        {
+            FactoryInfo.ProductionChain cChain = null;
```
You should avoid prefixes on variables. If you want it really short, maybe `prodChain`? I'd name it `productionChain`.

vogonistic commented on Assets/Scripts/Models/Furniture.Factory.cs:
```diff
@@ -0,0 +1,185 @@
+                foreach (var reqInputItem in cChain.Input)
+                {
+                    // check input slots for req. item:                        
+                    Tile tt = World.Current.GetTileAt(Tile.X + reqInputItem.SlotPosX, Tile.Y + reqInputItem.SlotPosY);
```
Avoid naming variables `tt` or similar. This should probably be `inputTile` or `inputSlotTile`.

vogonistic commented on Assets/Scripts/Models/Furniture.Factory.cs:
```diff
@@ -0,0 +1,185 @@
+
+
+            // TODO: can cause problems if there are more inputs and 1 is hanging
+            if (JobCount() > 0)
```
This should have brackets `{}` and probably a warning or error log if it shouldn't happen.

vogonistic commented on Assets/Scripts/Models/Furniture.cs:
```diff
@@ -20,7 +20,7 @@
 /// </summary>
 [MoonSharpUserData]
-public class Furniture : IXmlSerializable, ISelectable, IContextActionProvider
+public partial class Furniture : IXmlSerializable, ISelectable, IContextActionProvider
```
This project is about getting people involved with game programming, C# and Unity through a tile-based base building game. Quill has stayed away from several concept that he thought would be off-putting to new users and `partial` might very well be one of them. I don't know of it being used anywhere else.

In order to start using it, we probably need to discuss the potential value.

bjubes replied:
although its worth noting that by having `partial` people will learn what the keyword means and how splitting up classes can be useful in a large project.

sboigelot replied:
hem I'm not a fan either of partial, if you have to partial a class you can probably split it in two for separation of concerns.

vogonistic commented on Assets/StreamingAssets/Data/Furniture.xml:
```diff
@@ -277,6 +277,45 @@
+		<PossibleProductions>
+			<ProductionChain name="Iron smelting" processingTime="4">
+				<Input>
+					<Item objectType="Raw Iron" amount="3" slotPosX="0" slotPosY="0" />
```
`slotPosX` and `slotPosY` doesn't belong on each individual `Item`. It should probably be defined for the furniture as a whole.

Any particular reason it converts 3 for three? Is the idea that you can convert up to 48 iron for 48 steel? Or does it only take 3 items in the input?

dusho replied:
Ye, I had it like before, but this setup enables you to have different positions of output slots for different production chain and mainly this way, it will be kept always in sync - number of items in input and slots.. (like modder can specify 2 slots in input, but places actually 3 item types as requirement for production - this can cause some headache I wanted to prevent)

vogonistic replied:
Can you give me an example of a machine that would take input from one side for one conversion and somewhere else for another? I think it's unnecessary duplication of information. If you have a good example, but it's still a rare one, there should be a default value for the furniture and overrides other tags.

I like having the values to support conversion ratio, like it's specified on copper. (although, I'd use 1:2).

dusho replied:
example: you have one machine that creates some chips (or something), it has 2 production chains defined:
- crappy chip (requires 2 inputs - copper wire 1x, silicon wafer 1x)
- super cool chip (requires 3 inputs - copper wire 1x, silicon wafer 2x, diamond 1x)
Not sure I get the conversion ratios.. it's explicitly set by amounts on input and output (factorio style)

vogonistic replied:
We could also store the material inside the furniture, like in Factorio, and not have to bother with which side of the machine you input the material in. If we start having multiple inputs and store everything on different tiles, we are going to start having huge machines pretty quickly.

dusho replied:
sure.. but what does it mean 'inside furniture' cause you don't want to hide the input materials from the user.. it will be still placed next to machine (defined input slots)
also we are using fast little dudes to place things instead of fixed belts (like in Factorio), so having fixed input slots is ok, I think
flexible input gives you more choices and in this case prevents some setup errors, and needing to specify input slot position for every item isn't such a big extra
though it can still have default input and output slots defined for furniture.. have overrides (slotPosX and Y) as you mentioned .. 
It may be harder in have generic input slots and there may be some funky situations.. like you have two input slots, you need copper and steel.. 2 dudes will accidentally pick copper and try to feed the machine.. they will both place copper (using those 2 slots), then machine will not work.. you need to sync some job with furniture that you have to clean up one of the slots for steel..

vogonistic replied:
Factorial holds the material that has been fed to the machine inside it and you have to click on it to see how much it's holding. Same thing with material dropped of at a building site in Rimworld. I think that is a better solution. 

The reason I asked about 3 inputs is because we'd need 4 tiles to put and get material from plus the sprite of the machine, making them very big. 

It's a personal preference and something that I think is worth discussing separately.

vogonistic commented:
+1 on adding Copper, but that could probably have been added in a small separate PR which I'd taken in a heartbeat.

In general I like the consume/produce pattern. This looks like a pretty good improvement to the current way of doing conversions on top of that.

svmnotn commented:
@dusho this PR has some conflicts, please fix them

zwrawr commented on Assets/StreamingAssets/Data/Furniture.xml:
```diff
@@ -277,6 +277,45 @@
+
+    <JobSpotOffset X="1" Y="0" />
+
+	<FactoryInfo>
```
Your xml has a mix of spaces and tabs, pick one or the other :)

dusho commented:
made some changes.. still having problem with context menu Action not changing my params (some problem with being a copy or something and not updating real furniture params)
Added new temp art here
![image](https://cloud.githubusercontent.com/assets/12616499/18293689/cc091a74-7494-11e6-9163-e6f378273f97.png)

dusho commented:
new commit, got rid of that partial class.. factory overall renamed to workshop.. to have class FurnitureWorkshop (instead of FurnitureFactory that can be confused with design pattern)
still having issue with context menu.. last is triggered no matter which one of workshop's context items is selected

dusho commented:
merged with upstream
added copper plate and renamed raw_copper to Raw Copper for consistency with other resources
there seems to be funky behavior of jobs when items are not available (job is put into waitingJobQueue and then never put into active jobQueue, even if items are available) - not related to this commit

Grenkin1988 commented on Assets/StreamingAssets/Images/Inventory/Raw Copper.xml:
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Sprites>
+	<Sprite name="Raw Copper"     x="0" y="0" w="1" h="1" pixelPerUnit="64" />
+</Sprites>
```
Add one line at the end

Grenkin1988 commented on Assets/StreamingAssets/Images/Inventory/Power Cell.xml:
```diff
@@ -1,4 +1,4 @@
 <Sprites>
-	<Sprite name="Power Cell"     x="0" y="0" w="1" h="1" pixelPerUnit="128" />
+	<Sprite name="Power Cell"     x="0" y="0" w="1" h="1" pixelPerUnit="64" />
 </Sprites>
```
Add one line at the end

Grenkin1988 commented on Assets/StreamingAssets/Images/Inventory/Copper Plate.xml:
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Sprites>
+	<Sprite name="Copper Plate"     x="0" y="0" w="1" h="1" pixelPerUnit="64" />
+</Sprites>
```
Add one line at the end

Grenkin1988 commented on Assets/StreamingAssets/Images/Furniture/Metal Furnace.xml:
```diff
@@ -0,0 +1,9 @@
+	<Sprite name="Metal_Furnace_Running_3"     	x="6" y="3" w="3" h="3" pixelPerUnit="32" />
+	<Sprite name="Metal_Furnace_Running_4"     	x="9" y="3" w="3" h="3" pixelPerUnit="32" />
+	<Sprite name="Metal_Furnace_Running_5"     	x="12" y="3" w="3" h="3" pixelPerUnit="32" />
+</Sprites>
```
Add one line at the end

Grenkin1988 commented on Assets/Scripts/Models/FurnitureWorkshop.cs:
```diff
@@ -0,0 +1,343 @@
+[XmlRoot("Workshop")]
+public class FurnitureWorkshop
+{
+    #region Serialization
```
It is bad practice to use `region` to hide something

Grenkin1988 commented:
`FurnitureWorkshop` has StyleCop errors

Grenkin1988 commented on Assets/Scripts/Models/FurnitureWorkshop.cs:
```diff
@@ -0,0 +1,343 @@
+                        var oldProd = furn.Parameters[CUR_PRODUCTION_CHAIN_PARAM_NAME];
+                        var isNotProcessing = furn.Parameters[CUR_PROCESSED_INV_PARAM_NAME].ToInt() == 0;
+                        // if selected production really changes and nothing is being processed now
+                        if (!prod.Equals(oldProd) && isNotProcessing)
```
`!prod.Equals(oldProd.Value)` - this is more obvious.

Grenkin1988 commented on Assets/Scripts/Models/FurnitureWorkshop.cs:
```diff
@@ -0,0 +1,343 @@
+using System.Text;
+using System.Xml;
+using System.Xml.Serialization;
+using UnityEngine;
```
Remove unused directives

Grenkin1988 commented on Assets/Scripts/Models/FurnitureWorkshop.cs:
```diff
@@ -0,0 +1,343 @@
+
+        if (!string.IsNullOrEmpty(curSetupChainName))
+        {
+            FurnitureWorkshop.ProductionChain prodChain = GetProductionChainInfo(curSetupChainName);
```
This can be -> `ProductionChain prodChain = GetProductionChainInfo(curSetupChainName);`

Grenkin1988 commented on Assets/Scripts/Models/FurnitureWorkshop.cs:
```diff
@@ -0,0 +1,343 @@
+        foreach (var reqInputItem in prodChain.Input)
+        {
+            // if there is no hauling job for input object type, create one
+            var existingHaulingJob = furniture.jobs.Any(x => x.inventoryRequirements.ContainsKey(reqInputItem.ObjectType));
```
How about this:
```
Job furnitureJob;
string requiredType = reqInputItem.ObjectType;
bool existingHaulingJob = furniture.FindJobByPredicate(job => job.inventoryRequirements.ContainsKey(requiredType), out furnitureJob);
```
?

Aand add:
```
public bool FindJobByPredicate(Func<Job, bool> predicate, out Job job)
    {
    job = jobs.FirstOrDefault(predicate);
    return job != null;
}
```
to `Furniture`

dusho replied:
I think I incorporated all comments, but not sure about this one..
That method in Furniture would be linq returning object (FirstOrDefault) and then checking if it's null.. It's basically what Any does
Your proposed code is exactly (if not more) wordy than original one.
I think it's cool just to have collection and let people do appropriate linqs on that..

Grenkin1988 replied:
**Main** think is than `jobs` stays private in `Furniture`

dusho replied:
is there a reason to keep it private there? exposing collection enables linq queries for various needs (even for modders?). People will be adding their queries to furniture all the time.. is that desired?
is it that with getter on collections you are actually able to add items to it?
there is another PR #1201 that exposes only indexer to jobs

Grenkin1988 replied:
Exposing collection could lead to adding new elements or clearing it without appropriate logic.

dusho replied:
ye.. you were faster than my edit.. I agree that public getters on collection is asking for problems in C#

Grenkin1988 commented on Assets/Scripts/Models/FurnitureWorkshop.cs:
```diff
@@ -0,0 +1,343 @@
+
+    public string GetDescription()
+    {
+        StringBuilder sb = new StringBuilder();
```
Like this:
```
StringBuilder sb = new StringBuilder();
string prodChain = furnParameters[CUR_PRODUCTION_CHAIN_PARAM_NAME].ToString();
sb.AppendLine(!string.IsNullOrEmpty(prodChain) ? string.Format("Production: {0}", prodChain) : "No selected production");
return sb.ToString();
```
?

Grenkin1988 commented on Assets/Scripts/Models/FurnitureWorkshop.cs:
```diff
@@ -0,0 +1,343 @@
+                WorkshopMenuActions.Add(new WorkshopContextMenu()
+                {
+                    ProductionChainName = prodChainName,
+                    Function = (furn, prod) =>
```
I think more readable would be change lambda by simple method. ->
```
WorkshopMenuActions.Add(new WorkshopContextMenu
                {
                    ProductionChainName = chain.Name,
                    Function = ChangeCurrentProductionChain
                });
```
and
```
private static void ChangeCurrentProductionChain(Furniture furniture, string productName)
    {
        Parameter currentProdaction = furniture.Parameters[CUR_PRODUCTION_CHAIN_PARAM_NAME];
        bool isNotProcessing = furniture.Parameters[CUR_PROCESSED_INV_PARAM_NAME].ToInt() == 0;
        // if selected production really changes and nothing is being processed now
        if (productName.Equals(currentProdaction.Value) || !isNotProcessing)
        {
            return;
        }

        furniture.CancelJobs();
        furniture.Parameters[CUR_PRODUCTION_CHAIN_PARAM_NAME].SetValue(productName);
    }
```

Grenkin1988 commented on Assets/Scripts/Models/FurnitureWorkshop.cs:
```diff
@@ -0,0 +1,343 @@
+        return flaggedForTaking;
+    }
+
+    private FurnitureWorkshop.ProductionChain GetProductionChainInfo(string curSetupChainName)
```
->
```
private ProductionChain GetProductionChainByName(string productionChainName)
{
    return PossibleProductions.FirstOrDefault(chain => chain.Name.Equals(productionChainName));
}
```

dusho commented:
incorporated comments from Grenkin1988 (thanks for great review!)
fixed StyleCop warnings

Grenkin1988 commented on Assets/Scripts/Models/FurnitureWorkshop.cs:
```diff
@@ -279,65 +294,66 @@ private List<TileObjectTypeAmount> CheckForInventoryAtOutput(FurnitureWorkshop.P
+        [XmlAttribute("processingTime")]
+        public float ProcessingTime { get; set; }
 
-public class WorkshopContextMenu
```
Oops!

Grenkin1988 commented:
You removed `WorkshopContextMenu`

dusho commented:
sorry.. forgot to add newly created file. fixed

sboigelot commented:
This branch has conflicts that must be resolved

sboigelot commented:
This branch break something, the character do not find job anymore: do not happen on master:

```
FilippoLeon did not find a job.
UnityEngine.Debug:Log(Object)
UberLogger.Logger:ForwardToUnity(Object, LogSeverity, Object, Object[]) (at Assets/UberLogger/UberLogger.cs:504)
UberLogger.Logger:Log(String, Object, LogSeverity, Object, Object[]) (at Assets/UberLogger/UberLogger.cs:452)
Debug:ULogChannel(String, String, Object[]) (at Assets/UberLogger/Debug.cs:233)
Character:GetNewJob() (at Assets/Scripts/Models/Character.cs:608)
Character:Update_DoJob(Single) (at Assets/Scripts/Models/Character.cs:707)
Character:Update(Single) (at Assets/Scripts/Models/Character.cs:346)
World:TickEveryFrame(Single) (at Assets/Scripts/Models/World.cs:167)
WorldController:Update() (at Assets/Scripts/Controllers/WorldController.cs:162)

```

dusho commented:
can you describe the testcase? cause nothing different should happen for existing furniture (as it uses LUA code for update). But new code kicks in only for power cell press.. and metal furnace + there is trick with furnace that you have to pick what to smelt there in context menu (right click) , iron ore -> iron plate, copper ore -> copper plate

sboigelot commented:
@dusho, I just tried it again. The test case is: start the game, open the unity console and check the console :)

![image](https://cloud.githubusercontent.com/assets/6545998/18419775/1311d0c4-7863-11e6-94d2-869cf7f917bc.png)


The character stay stuck at the center tile and never does anything

sboigelot commented:
Sorry, my bad, I'm silly, we have to order them to mine now, they won't stockpile ore by themself. This is confusing

dusho commented:
well actually something is indeed broken.. in my branch they are not picking to ice at all.. in master it works
I'm confused, as I did not change stockpile lua code.. and changes in Furniture.cs are related only to Workshop code.. I don't get it.. feels like some automerge messed up something somewhere, but I can't tell where

sboigelot commented on Assets/StreamingAssets/Data/Inventory.xml:
```diff
@@ -1,9 +1,9 @@
 ﻿<?xml version="1.0" encoding="utf-8" ?>
 <Inventories>
     <Inventory type="Raw Iron" maxStackSize="50" basePrice="0.8" category="inv_cat_raw"/>
-    <Inventory type="Ice" maxStackSize="10" basePrice="1.2" category="inv_cat_raw"/>
```
this may be the issue about ice? just quick reading

dusho replied:
ye.. wish I saw your comment sooner :)

dusho commented:
ok fixed.. 	ice got removed from inventory.xml somehow
